### PR TITLE
Multiple sqlalchemy engines

### PR DESCRIPTION
**Pull request by ysimonson**
Adds support for multiple sqlalchemy engines. This allows you to maintain a connection to more than one database at a time.

### CODE REVIEW

**Comment by ian-axelrod** [comment]
Tests are failing. I wasn't aware oz was integrated with CCI?

**Comment by ysimonson** [comment]
It isn't. Someone must've mistakenly added CCI to oz.

**Comment by ian-axelrod** [comment]
I removed it now.

File: oz/sqlalchemy/__init__.py [hunk 1]
```diff
@@ -6,16 +6,15 @@
...
 import oz
 
 from .actions import *
 from .middleware import *
 from .options import *
 
 Base = declarative_base()
-_engine = None
-_session = None
+_engines = {}
```
Comment (ian-axelrod) [inline oz/sqlalchemy/__init__.py]: Think we could use `dict()` here to make it clear we are using a dictionary constructor and not an empty set constructor?

**Comment by ysimonson** [comment]
Heh, me too.

File: oz/sqlalchemy/__init__.py [hunk 2]
```diff
@@ -26,26 +25,29 @@ class _AfterFork(object):
...
+    return Base.metadata.create_all(engine(connection_string=connection_string))
 
-def engine():
-    global _engine
+def engine(connection_string=None):
+    global _engines
+    connection_string = connection_string or oz.settings["db"]
 
-    if _engine == None:
+    if not connection_string in _engines:
```
Comment (ian-axelrod) [inline oz/sqlalchemy/__init__.py]: `if connection_string not in _engines:`

File: oz/sqlalchemy/__init__.py [hunk 3]
```diff
@@ -55,17 +57,21 @@ def engine():
         if oz.settings["db_pool_timeout"]:
             kwargs["pool_timeout"] = oz.settings["db_pool_timeout"]
 
-        _engine = create_engine(oz.settings["db"], **kwargs)
-        after_fork.registered = True
-        register_after_fork(after_fork, after_fork)
+        first_engine = len(_engines) == 0
+        _engines[connection_string] = create_engine(connection_string, **kwargs)
+
+        if first_engine:
```
Comment (ian-axelrod) [inline oz/sqlalchemy/__init__.py]: `if not _engines:`

File: oz/sqlalchemy/middleware.py [hunk 4]
```diff
@@ -17,42 +17,57 @@ def __init__(self):
...
             except:
                 tornado.log.app_log.warning("Error occurred during database transaction cleanup: %s", str(sys.exc_info()[0]))
                 raise
             finally:
-                self.db_conn.close()
+                for db_conn in self._db_conns.values():
+                    try:
+                        db_conn.close()
+                    except:
+                        pass
```
Comment (ian-axelrod) [inline oz/sqlalchemy/middleware.py]: We should log the exception, in the rare case it occurs.

File: oz/sqlalchemy/middleware.py [hunk 5]
```diff
@@ -17,42 +17,57 @@ def __init__(self):
...
```
Comment (ian-axelrod) [inline oz/sqlalchemy/middleware.py]: We should not the exception, in the rare case it occurs.

File: oz/sqlalchemy/__init__.py (hunk 3 above)
Comment (ysimonson) [inline oz/sqlalchemy/__init__.py]: That's not the same

Comment (ian-axelrod) [inline oz/sqlalchemy/__init__.py]: ?? _engines is a `dict`, right? If so `if not _engines` is the shorthand for checking that `_engines` is an empty dictionary.

Comment (ysimonson) [inline oz/sqlalchemy/__init__.py]: It's never empty by the time it gets to this line (see line 61.)

**Comment by ian-axelrod** [comment]
lgtm

**Pull request merged**